File: src/legacy/SxxMachine/builtin/PRED_current_input_1.cpp
using namespace std;

#include "PRED_current_input_1.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../machine/SxxMachine/VariableTerm.h"
#include "../../../machine/SxxMachine/FFIObjectTerm.h"
#include "../../../exceptions/SxxMachine/IllegalDomainException.h"

namespace SxxMachine
{
	using FFIObjectTerm = SxxMachine::FFIObjectTerm;
	using Operation = SxxMachine::Operation;
	using Predicate = SxxMachine::Predicate;
	using Prolog = SxxMachine::Prolog;
	using Term = SxxMachine::Term;
	using VariableTerm = SxxMachine::VariableTerm;
	using IllegalDomainException = SxxMachine::IllegalDomainException;

	PRED_current_input_1::PRED_current_input_1(Term *a1, Operation cont)
	{
	LARG[0] = a1;
	this->cont = cont;
	}

	Operation PRED_current_input_1::exec(Prolog *engine)
	{
		engine->requireFeature(Prolog::Feature::IO, this, LARG[0]);
		engine->setB0();
	Term *a1;
	a1 = LARG[0];
	a1 = a1->dref();
	if ((dynamic_cast<VariableTerm*>(a1) != nullptr))
	{
		FFIObjectTerm tempVar(engine->getCurrentInput());
		a1->asVariablelTerm().bind(&tempVar, engine->trail);
	}
	else if ((dynamic_cast<FFIObjectTerm*>(a1) != nullptr))
	{
		FFIObjectTerm tempVar2(engine->getCurrentInput());
		if (!a1->unify(&tempVar2, engine->trail))
		{
		return engine->fail();
		}
	}
	else
	{
		throw IllegalDomainException(this,1,"stream",a1);
	}
	return cont;
	}
}

File: jsrc/legacy/SxxMachine/builtin/PRED_current_input_1.java
package SxxMachine.builtin;

import SxxMachine.FFIObjectTerm;
import SxxMachine.Operation;
import SxxMachine.Predicate;
import SxxMachine.Prolog;
import SxxMachine.Term;
import SxxMachine.VariableTerm;
import SxxMachine.IllegalDomainException;
/**
 * <code>current_input/1</code><br>
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 */
public class PRED_current_input_1 extends Predicate.P1 {
    public PRED_current_input_1(Term a1, Operation cont) {
	LARG[0] = a1;
	this.cont = cont;
    }

    public Operation exec(Prolog engine) {
        engine.requireFeature(Prolog.Feature.IO, this, LARG[0]);
        engine.setB0();
	Term a1;
	a1 = LARG[0];
	a1 = a1.dref();
	if ((a1 instanceof VariableTerm)) {
	    a1.asVariablelTerm().bind(new FFIObjectTerm(engine.getCurrentInput()), engine.trail);
	} else if ((a1 instanceof FFIObjectTerm)) {
	    if (! a1.unify(new FFIObjectTerm(engine.getCurrentInput()), engine.trail)) 
		return engine.fail();
	} else {
	    throw new IllegalDomainException(this,1,"stream",a1);
	}
	return cont;
    }
}
